### DISCUSSION TITLE
Self Hosted Project supabaseKey

### DISCUSSION THREAD

**Discussion by alexstaroselsky** [supabase/supabase#1029] [Questions] [answered]
For a [self-hosted project](https://supabase.io/docs/guides/self-hosting) created by supabase/cli `eject` to work with clients such as `supabase-js`, where/how should the `supabaseKey` be generated/specified? I see in the `./docker/kong/kong.yml` there is a section:

```yaml
consumers:
  - username: 'private-key'
    keyauth_credentials:
      - key: 
```

Is `consumers[0].keyauth_credentials[0].key` where you can specify any string you want to be the `supebaseKey` to be provided to an instance of a `supabase-js` client? As in:

```js
import { createClient } from '@supabase/supabase-js'

// Create a single supabase client for interacting with your database 
const supabase = createClient("http://localhost:8000", "where is this generated/specified for self hosted ejected project")
```

Or do you still create a project in the supabase dashboard, then use that api key generated for you as the `supabaseKey`, but still point the client to `localhost`?

Thanks!

**Comment by chetan** [comment]
@alexstaroselsky I ran into this as well and I think I found the solution (or workaround). Instead of `supabase eject` as the docs recommend, do this instead - 

```sh
supabase init
supabase start
```

`init` creates almost the same set of compose files but also populates the key and a couple of other things. Fairly sure that `start` is just calling `docker-compose up -d`.

**Comment by soedirgo** [comment, marked as answer]
You're right, the one in kong.yml is the anon key. I should've updated `eject` to print the keys as well, will update it in a bit.

(In case you're wondering, the Supabase key [just stores the Postgres role](https://jwt.io/#debugger-io?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJzdXBhYmFzZSIsImlhdCI6MTYwMzk2ODgzNCwiZXhwIjoyNTUwNjUzNjM0LCJyb2xlIjoiYW5vbiJ9.36fUebxgx1mcBo4s19v0SzqmzunP--hm_hep0uLX0ew) used when connecting. It's JWT with hardcoded secret of `super-secret-jwt-token-with-at-least-32-characters-long` right now, but I might change this soon, let the user specify it or some such.)

For now the workaround mentioned by @chetan works, all you need is the `docker` directory in `.supabase`. And yes, `supabase start` just uses `docker-compose` :). (I might need to look into [dockerode](https://github.com/apocas/dockerode).)

**Reply by alexstaroselsky** [reply]
Thank you, `init` and `start` generated the key. I do get `401` "Invalid authentication credentials" with the examples when using `http://localhost:8000` as the supabase url and either `super-secret-jwt-token-with-at-least-32-characters-long` (or the key in the kong.yml) as the supabase key. I guess what should environment values be for the supabase client when using the docker-compose from `supabase init`?

**Reply by chetan** [reply]
`docker-compose` is a python app which actually does quite a bit of heavy lifting. Not sure it would be worthwhile to reimplement all of that. That said, there is a proper [spec](https://compose-spec.io/) now, so there may be some other implementations out there.

**Reply by soedirgo** [reply]
@alexstaroselsky my bad, the `super-secret-...` is the JWT secret, the token is [here](https://github.com/supabase/cli/blob/main/src/commands/init.ts#L53-L56).

> docker-compose is a python app

TIL! I thought it was part of the `docker` CLI, or at least written in Go. There is the [Docker Engine API](https://docs.docker.com/engine/api/sdk/examples/) though, so it might not be necessary to use Compose.

**Reply by alexstaroselsky** [reply]
I see that output in the newest version of the supabase/cli. I still get 401 "Invalid authentication credentials" for sign-up and magiclink actions in projects such as `examples/nextjs-todo-list` using environment values in `.env.local`, similarly to the other examples:

```
NEXT_PUBLIC_SUPABASE_URL=http://localhost:8000
NEXT_PUBLIC_SUPABASE_ANON_KEY=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJzdXBhYmFzZSIsImlhdCI6MTYwMzk2ODgzNCwiZXhwIjoyNTUwNjUzNjM0LCJyb2xlIjoiYW5vbiJ9.36fUebxgx1mcBo4s19v0SzqmzunP--hm_hep0uLX0ew
```

Log output of `supabase-kong` being for operations like `POST /auth/v1/signup` and `POST /auth/v1/magiclink`:

```
supabase-kong | someip - - [02/Apr/2021:16:29:40 +0000] "OPTIONS /auth/v1/signup HTTP/1.1" 200 0 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/anotherip Safari/537.36"
supabase-kong | someip - - [02/Apr/2021:16:29:40 +0000] "POST /auth/v1/signup HTTP/1.1" 401 52 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/anotherip Safari/537.36"
supabase-kong | someip - - [02/Apr/2021:16:30:12 +0000] "OPTIONS /auth/v1/magiclink HTTP/1.1" 200 0 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/anotherip Safari/537.36"
supabase-kong | someip - - [02/Apr/2021:16:30:12 +0000] "POST /auth/v1/magiclink HTTP/1.1" 401 52 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/anotherip Safari/537.36"
```

What's interesting to me is that it doesn't seem to hit `supabase-auth` when that 401 happens, so it's stopping/failing at the kong gateway.

**Reply by soedirgo** [reply]
Can you try [this one](https://github.com/supabase/cli/blob/main/src/commands/eject.ts#L36-L37)? Looks like I used a different anon key for `eject` - will make it match the ones in `init` later.

**Reply by soedirgo** [reply]
Scratch that, I realized what the problem was: the `key` field was empty. This should have been populated during `init`/`eject`. I'll push a fix real quick.

**Reply by soedirgo** [reply]
Should be fixed in v0.3.6. Tested with the nextjs-todo-list example with the docker build from `eject`.

**Reply by alexstaroselsky** [reply]
Newest version worked great with command `eject`. Was able to use the key with the various examples as well as authenticate. Thanks!
